src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala

```diff
@@ -473,30 +467,55 @@ class QueryCompiler(cq : ConjunctiveQuery) {
   // generate inference rule part for deepdive
   def compile(stmt: InferenceRule): CompiledBlocks = {
     val headAsBody = stmt.head.variables map (_.atom)
-    val variableIdAndKeyColumns = headAsBody.zipWithIndex flatMap {
+    val variableIdAndColumns = headAsBody.zipWithIndex flatMap {
       case (x: Atom, i) if schemaDeclarationByRelationName get x.name exists (_.isQuery) =>
         // TODO maybe TableAlias can be useful here or we can completely get rid of it?
         // variable id column
-        s"""R${headAsBody indexOf x}.${
+        s"""R_${headAsBody indexOf x}.${
           deepdiveVariableIdColumn
-        } AS "${x.name}.R${headAsBody indexOf x}.${deepdiveVariableIdColumn}\"""" :: (
+        } AS "${x.name}.R${headAsBody indexOf x}.${deepdiveVariableIdColumn}\"""" ::
+        List(
           // project variable key columns as well (to reduce unnecssary joins)
           schemaDeclarationByRelationName get x.name map (_.keyColumns map {
             case term => s"""R${headAsBody indexOf x}.${term
             } AS "${x.name}.R${headAsBody indexOf x}.${term}\""""
           }) get
-        )
+        ).flatten ::: List(
+          // project category value columns as well (to reduce unnecssary joins)
+          schemaDeclarationByRelationName get x.name map (_.categoricalColumns map {
+            case term => s"""R${headAsBody indexOf x}.${term
+            } AS "${x.name}.R${headAsBody indexOf x}.${term}\""""
+          }) get
+        ).flatten
+
+      case _ => List.empty
+    }
+
+    val internalVarTables = headAsBody.zipWithIndex flatMap {
+      case (x: Atom, i) if schemaDeclarationByRelationName get x.name exists (_.isQuery) =>
+        List(s"""${deepdivePrefixForVariablesIdsTable}${x.name} AS R_${headAsBody indexOf x}""")
       case _ => List.empty
     }
 
+    val internalVarJoinConds = headAsBody.zipWithIndex flatMap {
+      case (x: Atom, i) if schemaDeclarationByRelationName get x.name exists (_.isQuery) =>
+        List(
+          // project variable key columns as well (to reduce unnecssary joins)
+          schemaDeclarationByRelationName get x.name map (_.keyColumns map {
+            case term => s"""R${headAsBody indexOf x}.${term} = R_${headAsBody indexOf x}.${term}"""
+          }) get
+        )
+      case _ => List.empty
+    } flatten
+
     val inputQueries =
       stmt.q.bodies map { case cqBody =>
         // Here we need to select from the bodies atoms as well as the head atoms,
         // which are the variable relations.
         // This is achieved by puting head atoms into the body.
-        val fakeBody        = (headAsBody map { h => h.copy(name = s"$deepdivePrefixForVariablesIdsTable${h.name}") }) ++ cqBody
+        val fakeBody        = headAsBody ++ cqBody
         val fakeCQ          = stmt.q.copy(bodies = List(fakeBody))
-        val qc = new QueryCompiler(fakeCQ)
+        val qc = new QueryCompiler(fakeCQ, internalVarTables, internalVarJoinConds)
```

Comment:
Right, we could add a routine to generate these extra stuff for each head or (fake head) atom. Leaving that to another day...